### DISCUSSION TITLE
bounds, AABB and AABBLike

### DISCUSSION THREAD

**Discussion by dennemark** [thi-ng/umbrella#455] [Q&A] [answered]
Hi,
I want to do bounding box intersection check.
I am providing `bounds` some points `bounds(points3(myPoints))` but it returns an AABBLike.
If i want to use the result within `intersectionAABB` typescript complains that it is not an `AABB` type.
What is the purpose of AABBLike? Couldn`t we just return an AABB from bounds?
Right now i am doing the following: `aabb(bounds(....))`

Or is the main reason that I have to decide afterwards, if I want to have a rect (2D) or an aabb (3d) ?

Kind regards!

**Comment by postspectacular** [comment, marked as answer]
Hi Martin - the purpose of the `AABBLike` interface is a generalization for both 2D/3D cases and it's currently implemented by both the `Rect` (for 2D) and `AABB` (for 3D) classes. The `bounds()` function is polymorphic and dynamically chooses the right impl based on given inputs, but TypeScript only allows to specify a single return type (hence the `AABBLike` to satisfy both dimensions). In your case, `bounds()` will delegate to an implementation for `points3` and does actually return an `AABB` shape instance, but again, the `bounds()` function's return type is more general...

So to use the result with `intersectionAABB()` you just need to add a manual cast to `<AABB>`, but **do not** require a separate coercion via `aabb(...)`, i.e. you only need this:

```ts
intersectionAABB(
  <AABB>bounds(points3(myPoints)),
  <AABB>bounds(points3(myOtherPoints))
);
```

I should update the `bounds()` docs to better clarify what is being returned, i.e. a `Rect` instance for all 2D shapes and an `AABB` for 3D cases... There're some cases (e.g. empty groups) where the result can also be `undefined`...

**Reply by dennemark** [reply]
Ah this is convenient! I thought bounds might only return an interface, without functionality. But now it makes sense!
I am using typescript for a while now, but wasn't aware that this kind of casting  is possible, too `<AABB>bounds`.

**Reply by postspectacular** [reply]
Just to add here: An interface/type only exists during authoring/compiling. It's best to think of them as contracts. A value/object either satisfies the contract or not. At runtime there're only values (no more types). Class instances in JS are just objects and to be compatible with an interface, they only must conform to the shape/structure of that interface/type (hence TypeScript is also known as structurally typed language, in contrast to say Java or C++, which are statically typed, with more limitations)...

So when we say here that `bounds()` has an `AABBLike` return type, you should read it as a value which satisfies the contract of the `AABBLike` interface, but provides no other promises or known features:

https://github.com/thi-ng/umbrella/blob/6e20f80dd9df1c8055ffa3c1e4d6f7598add0c0b/packages/geom-api/src/shape.ts#L18-L24
